pkg/api/serve.go

```diff
@@ -71,25 +68,70 @@ func Serve(
 		actions,
 		logger,
 	)
-	controller.Configure(api)
+	HandlerFromMuxWithBaseURL(controller, apiRouter, BaseURL)
 
-	apiHandler := api.Serve(func(handler http.Handler) http.Handler {
-		// build handler for our REST API
-		return httputil.LoggingMiddleware(
-			RequestIDHeaderName,
-			logging.Fields{"service_name": LoggerServiceName},
-			promhttp.InstrumentHandlerCounter(requestCounter,
-				MetricsHandler(api.Context(),
-					NewCookieAPIHandler(handler))))
-	})
 	uiHandler := NewUIHandler(authService, gatewayDomain)
+	r.Mount("/_health", httputil.ServeHealth())
+	r.Mount("/metrics", promhttp.Handler())
+	r.Mount("/_pprof/", httputil.ServePPROF("/_pprof/"))
+	r.Mount("/", uiHandler)
+	return r
+}
+
+// OapiRequestValidatorWithOptions Creates middleware to validate request by swagger spec.
+// This middleware is good for net/http either since go-chi is 100% compatible with net/http.
+func OapiRequestValidatorWithOptions(swagger *openapi3.Swagger, options *openapi3filter.Options) func(http.Handler) http.Handler {
+	router, err := legacy.NewRouter(swagger)
+	if err != nil {
+		panic(err)
+	}
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// validate request
+			if statusCode, err := validateRequest(r, router, options); err != nil {
+				http.Error(w, err.Error(), statusCode)
+				return
+			}
+			// serve
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func validateRequest(r *http.Request, router routers.Router, options *openapi3filter.Options) (int, error) {
+	// Find route
+	route, pathParams, err := router.FindRoute(r)
+	if err != nil {
+		return http.StatusBadRequest, err // We failed to find a matching route for the request.
+	}
+
+	// Validate request
+	requestValidationInput := &openapi3filter.RequestValidationInput{
+		Request:    r,
+		PathParams: pathParams,
+		Route:      route,
+		Options:    options,
+	}
+
+	if _, ok := route.Operation.Extensions["x-validation-exclude-body"]; ok {
+		o := *options
+		o.ExcludeRequestBody = true
+		requestValidationInput.Options = &o
+	}
+
+	if err := openapi3filter.ValidateRequest(r.Context(), requestValidationInput); err != nil {
+		var reqErr *openapi3filter.RequestError
+		if errors.As(err, &reqErr) {
+			return http.StatusBadRequest, err
+		}
+		var seqErr *openapi3filter.SecurityRequirementsError
```

Comment:
Who returns this error? I thought we're in charge of auth and we don't return this type.